codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/KotlinSettings.kt

```diff
@@ -263,16 +263,20 @@ enum class DefaultValueSerializationMode(val value: String) {
  * @param visibility Enum representing the visibility of code-generated classes, objects, interfaces, etc.
  * @param nullabilityCheckMode Enum representing the nullability check mode to use
  * @param defaultValueSerializationMode Enum representing when default values should be serialized
+ * @param enableEndpointAuthProvider flag indicating that endpoint resolution should be enabled as part of resolving
+ * an auth scheme. This is an advanced option that only a select few service clients like S3 and EventBridge require.
  */
 data class ApiSettings(
     val visibility: Visibility = Visibility.PUBLIC,
     val nullabilityCheckMode: CheckMode = CheckMode.CLIENT_CAREFUL,
     val defaultValueSerializationMode: DefaultValueSerializationMode = DefaultValueSerializationMode.WHEN_DIFFERENT,
+    val enableEndpointAuthProvider: Boolean = false,
 ) {
```

Comment:
This is hard to answer succinctly. There has been a ton of discussion and docs written around this. The only two services that need this right now are S3 and EventBridge. Using endpoints for auth resolution is an anti-pattern and we should be finding other ways going forward. The issue is those services already shipped and it's currently the only way to get Sigv4A support there because it's conditionally enabled by input members being set to specific things.

Turning this on should ultimately not be a breaking change. The only supported auth schemes that should show up are ones supported by the model otherwise we'd have a SEP on it. If it puts your mind at ease Java did the [same](https://github.com/aws/aws-sdk-java-v2/blob/master/services/s3/src/main/resources/codegen-resources/customization.config#L264C4-L264C34) thing.

Reply:
Aren't generated `AuthSchemeParameters` interfaces public? Turning this setting on causes an extra `endpointParameters` member to be added to the interface which sounds backwards incompatible.

Reply:
Endpoints aside we should expect over time there may be additional parameters added to either endpoint or auth scheme resolver parameters. I'm not sure how this is different. We would have to default the values I suppose but I don't think it would be breaking or backwards incompatible.